File: solutions/python/0345-reverse-vowels-of-a-string.py
class Solution:
    def reverseVowels(self, s: str) -> str:
        vowels = set("aeiouAEIOU") 
        s = list(s) 
        start, end = 0, len(s) - 1 
        while start < end: 
            if s[start] in vowels and s[end] in vowels: 
                s[start], s[end] = s[end], s[start] 
                start += 1 
                end -= 1 
            elif s[start] not in vowels: 
                start += 1 
            elif s[end] not in vowels: 
                end -= 1 
        return "".join(s) 
                                                                                                                                                                       

File: solutions/cpp/0345-reverse-vowels-of-a-string.cpp
class Solution {
public:
    string reverseVowels(string s) {
        int start = 0, end = s.size() - 1;
        unordered_set<char> vowels = {'a', 'e', 'i', 'o', 'u', 'A', 'E', 'I', 'O', 'U'};
        
        while (start < end) {
            if (vowels.count(s[start]) && vowels.count(s[end])) {
                swap(s[start++], s[end--]);
            } else if (vowels.count(s[start])) {
                end--;
            } else {
                start++;
            }
        }
        return s;
    }
};
